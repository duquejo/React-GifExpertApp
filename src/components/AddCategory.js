import { useState } from "react";
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

  const [ inputValue, setInputValue] = useState(''); // ''

  const handleInputChange = ( e ) => {
    setInputValue( e.target.value ); // Setting Input event
  };

  const handleSubmit = ( e ) => {
    
    e.preventDefault();
    // Greater than 2 value

    if( inputValue.trim().length > 2 ) {
      // setCategories( [...categories, 'Fullmetal Alchemist'] );
      setCategories( cats => [ inputValue, ...cats ] ); // Callback with the last state
      setInputValue( '' );
    }
  };

  return (
    <form onSubmit={ handleSubmit }>
      <input 
        type="text"
        value={ inputValue }
        onChange={ handleInputChange }
        placeholder="Type something and press 'enter' for giphy search!"
      />
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
};