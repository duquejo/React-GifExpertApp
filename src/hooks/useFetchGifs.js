/**
 * 
 * use 'prefix' is for hooks indication.
 */

import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = ( category ) => {

  /**
   * useEffect updates the state conditionally
   * useState updates all the component
   */

  const [state, setState] = useState({
    data: [],
    loading: true
  });

  /**
   * The effects always expects to be synchronous
   */
  useEffect( () => {

    getGifs( category ).then( imgs => {

      setState({
        data: imgs,
        loading: false
      });
        
    });

  },  [ category ] ); // If the dependencies list is empty, It's only fired ONCE (ComponentDidMount)

  return state; // { data:[], loading: true }
};