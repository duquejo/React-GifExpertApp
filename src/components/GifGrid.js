
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from 'prop-types';

export const GifGrid = ( { category } ) => {

  const { data:images, loading } = useFetchGifs( category );

  return (
    <>
      <h3 className="animate__animated animate__fadeIn"> { category } </h3>
      { loading && <p>Loading...</p> }
      <div className="card-grid">
        {
          images.map( img => <GifGridItem key={ img.id } { ...img } /> ) // Sending all properties with Object Rest operator
        }
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
};