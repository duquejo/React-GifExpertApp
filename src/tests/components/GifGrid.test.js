import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock( "../../hooks/useFetchGifs" );

describe('<GifGrid /> unit tests', () => {
  
  const category = 'Dragon Ball Z'; 


  test('Should render a <GifGrid /> component successfully', () => {

    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    });

    const wrapper  = shallow( <GifGrid category={ category }/> );    
    expect( wrapper ).toMatchSnapshot();
  });
  
  test('Should return items when the images are loaded useFetchGifs', () => {
    
    const gifs = [{
      id: '12345',
      url: 'https://127.0.0.1/something.gif',
      title: 'Someting'
    },{
      id: '67890',
      url: 'https://127.0.0.1/something2.gif',
      title: 'Someting 2'
    }];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false
    });

    const wrapper = shallow( <GifGrid category={ category }/> );

    expect( wrapper ).toMatchSnapshot();
    
    expect( wrapper.find('p') ).not.toBe(); // Assert if the p (loading) doesn't exists

    expect( wrapper.find('GifGridItem').length ).toBe( gifs.length ); // Assert the gifs length rendered
  });
});