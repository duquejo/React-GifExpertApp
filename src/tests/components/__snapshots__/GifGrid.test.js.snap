// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GifGrid /> unit tests Should render a <GifGrid /> component successfully 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Dragon Ball Z
     
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`<GifGrid /> unit tests Should return items when the images are loaded useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Dragon Ball Z
     
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="12345"
      key="12345"
      title="Someting"
      url="https://127.0.0.1/something.gif"
    />
    <GifGridItem
      id="67890"
      key="67890"
      title="Someting 2"
      url="https://127.0.0.1/something2.gif"
    />
  </div>
</Fragment>
`;
